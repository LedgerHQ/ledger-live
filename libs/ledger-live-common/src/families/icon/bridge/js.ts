import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";

import { sync, scanAccounts } from "../js-synchronization";
import { createTransaction } from "../js-createTransaction";
import { prepareTransaction } from "../js-prepareTransaction";
import getTransactionStatus from "../js-getTransactionStatus";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import { defaultUpdateTransaction } from "@ledgerhq/coin-framework/bridge/jsHelpers";

const receive = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
  preload: () => Promise.resolve({}),
  hydrate: () => {},
  scanAccounts,
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction: defaultUpdateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export default { currencyBridge, accountBridge };
