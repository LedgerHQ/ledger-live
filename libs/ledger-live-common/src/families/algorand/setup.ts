// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
