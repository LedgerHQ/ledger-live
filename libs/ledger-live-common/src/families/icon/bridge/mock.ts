import { BigNumber } from "bignumber.js";
import { NotEnoughBalance, RecipientRequired, InvalidAddress, FeeTooHigh } from "@ledgerhq/errors";
import type { Transaction } from "../types";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import { defaultUpdateTransaction } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import {
  scanAccounts,
  signOperation,
  broadcast,
  sync,
  isInvalidRecipient,
} from "../../../bridge/mockHelpers";
import { getMainAccount } from "../../../account";

import { makeAccountBridgeReceive } from "../../../bridge/mockHelpers";

const receive = makeAccountBridgeReceive();

const createTransaction = (): Transaction => ({
  family: "icon",
  mode: "send",
  amount: new BigNumber(0),
  recipient: "",
  useAllAmount: false,
  fees: null,
});

const updateTransaction = defaultUpdateTransaction;

const prepareTransaction = async (a, t) => t;

const estimateMaxSpendable = ({ account, parentAccount, transaction }) => {
  const mainAccount = getMainAccount(account, parentAccount);
  const estimatedFees = transaction?.fees || new BigNumber(5000);
  return Promise.resolve(BigNumber.max(0, mainAccount.spendableBalance.minus(estimatedFees)));
};

const getTransactionStatus = (account, t) => {
  const errors: any = {};
  const warnings: any = {};
  const useAllAmount = !!t.useAllAmount;

  const estimatedFees = new BigNumber(5000);

  const totalSpent = useAllAmount ? account.balance : new BigNumber(t.amount).plus(estimatedFees);

  const amount = useAllAmount ? account.balance.minus(estimatedFees) : new BigNumber(t.amount);

  if (amount.gt(0) && estimatedFees.times(10).gt(amount)) {
    warnings.amount = new FeeTooHigh();
  }

  if (totalSpent.gt(account.balance)) {
    errors.amount = new NotEnoughBalance();
  }

  if (!t.recipient) {
    errors.recipient = new RecipientRequired();
  } else if (isInvalidRecipient(t.recipient)) {
    errors.recipient = new InvalidAddress();
  }

  return Promise.resolve({
    errors,
    warnings,
    estimatedFees,
    amount,
    totalSpent,
  });
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

const currencyBridge: CurrencyBridge = {
  scanAccounts,
  preload: (async () => {}) as any,
  hydrate: () => {},
};

export default { currencyBridge, accountBridge };
