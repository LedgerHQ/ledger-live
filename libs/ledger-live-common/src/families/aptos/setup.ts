// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import {
  AptosAccount,
  TransactionStatus,
  createBridges,
  type Transaction,
} from "@ledgerhq/coin-aptos";
import Transport from "@ledgerhq/hw-transport";
import Aptos from "@ledgerhq/hw-app-aptos";
import type { Bridge } from "@ledgerhq/types-live";
import aptosResolver from "@ledgerhq/coin-aptos/signer/index";
import makeCliTools from "@ledgerhq/coin-aptos/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Aptos> = (transport: Transport) => {
  return new Aptos(transport);
};

const bridge: Bridge<Transaction, AptosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const resolver: Resolver = createResolver(createSigner, aptosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
