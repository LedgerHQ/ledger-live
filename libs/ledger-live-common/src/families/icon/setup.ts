// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-icon/bridge/index";
import makeCliTools from "@ledgerhq/coin-icon/cli-transaction";
import iconResolver from "@ledgerhq/coin-icon/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-icon/types/index";
import Icon from "@ledgerhq/hw-app-icon";
import Transport from "@ledgerhq/hw-transport";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";
import { IconCoinConfig } from "@ledgerhq/coin-icon/config";
import { getEnv } from "@ledgerhq/live-env";

const createSigner: CreateSigner<Icon> = (transport: Transport) => {
  return new Icon(transport);
};

const getCurrencyConfig = (): IconCoinConfig => {
  return {
    status: {
      type: "active",
    },
    infra: {
      indexer: getEnv("ICON_INDEXER_ENDPOINT"),
      indexer_testnet: getEnv("ICON_TESTNET_INDEXER_ENDPOINT"),
      node_endpoint: getEnv("ICON_NODE_ENDPOINT"),
      node_testnet_endpoint: getEnv("ICON_TESTNET_NODE_ENDPOINT"),
      debug_endpoint: getEnv("ICON_DEBUG_ENDPOINT"),
      debug_testnet_endpoint: getEnv("ICON_TESTNET_DEBUG_ENDPOINT"),
    },
  };
};

const bridge: Bridge<Transaction> = createBridges(
  executeWithSigner(createSigner),
  getCurrencyConfig,
);

const resolver: Resolver = createResolver(createSigner, iconResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
