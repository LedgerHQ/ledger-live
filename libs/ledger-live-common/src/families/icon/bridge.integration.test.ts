import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import { dataset } from "@ledgerhq/coin-icon/bridge.integration.test";

testBridge(dataset);
