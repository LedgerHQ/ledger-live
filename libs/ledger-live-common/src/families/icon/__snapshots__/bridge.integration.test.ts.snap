// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icon currency bridge scanAccounts icon seed 1 1`] = `
[
  {
    "balance": "798690000000000000",
    "currencyId": "icon",
    "derivationMode": "icon",
    "freshAddress": "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
    "freshAddressPath": "44'/4801368'/0'/0'/0'",
    "iconResources": {
      "nonce": 0,
      "totalDelegated": "0",
      "votingPower": "0",
    },
    "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
    "index": 0,
    "operationsCount": 24,
    "pendingOperations": [],
    "seedIdentifier": "0426dd3cdd69f3604e9c1d8f3021b20e133524f12894b9ef8e35d3f1f5244516d18b7bb0d1fee7b7c2972d5cde08190d66cf932d0addd5ac87f53198d050940ccb",
    "spendableBalance": "798690000000000000",
    "swapHistory": [],
    "syncHash": undefined,
    "used": true,
  },
  {
    "balance": "60000000000000",
    "currencyId": "icon",
    "derivationMode": "icon",
    "freshAddress": "hxfc049b4ce60cc5968487d180d9c4263e63181739",
    "freshAddressPath": "44'/4801368'/0'/0'/1'",
    "iconResources": {
      "nonce": 0,
      "totalDelegated": "0",
      "votingPower": "0",
    },
    "id": "js:2:icon:hxfc049b4ce60cc5968487d180d9c4263e63181739:icon",
    "index": 1,
    "operationsCount": 1,
    "pendingOperations": [],
    "seedIdentifier": "0426dd3cdd69f3604e9c1d8f3021b20e133524f12894b9ef8e35d3f1f5244516d18b7bb0d1fee7b7c2972d5cde08190d66cf932d0addd5ac87f53198d050940ccb",
    "spendableBalance": "60000000000000",
    "swapHistory": [],
    "syncHash": undefined,
    "used": true,
  },
  {
    "balance": "0",
    "currencyId": "icon",
    "derivationMode": "icon",
    "freshAddress": "hx0199324ab40847315ec3972d1781b9dccd1620ae",
    "freshAddressPath": "44'/4801368'/0'/0'/2'",
    "iconResources": {
      "nonce": 0,
      "totalDelegated": "0",
      "votingPower": "0",
    },
    "id": "js:2:icon:hx0199324ab40847315ec3972d1781b9dccd1620ae:icon",
    "index": 2,
    "operationsCount": 0,
    "pendingOperations": [],
    "seedIdentifier": "0426dd3cdd69f3604e9c1d8f3021b20e133524f12894b9ef8e35d3f1f5244516d18b7bb0d1fee7b7c2972d5cde08190d66cf932d0addd5ac87f53198d050940ccb",
    "spendableBalance": "0",
    "swapHistory": [],
    "syncHash": undefined,
    "used": false,
  },
]
`;

exports[`icon currency bridge scanAccounts icon seed 1 2`] = `
[
  [
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 81833326,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x09e597274b3dbf7e7225b248724f65e30a32a2c312a587e9999b7ff942585c2e",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x09e597274b3dbf7e7225b248724f65e30a32a2c312a587e9999b7ff942585c2e-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "750000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 80919307,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x0c3170fb250e6735e1a5f51dde9104fca3368ec745f6aa1e19679a9cc6d756da",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x0c3170fb250e6735e1a5f51dde9104fca3368ec745f6aa1e19679a9cc6d756da-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "1234000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 82738449,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x0f6a15de41871086480ebe1f42365df8e20537dc9eceae250bc91cbe29c017ca",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x0f6a15de41871086480ebe1f42365df8e20537dc9eceae250bc91cbe29c017ca-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "1200000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 80823008,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x13190d4a84cb2e088b858da3aa4f66e99e5c8444831e1701220142fa33640f41",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x13190d4a84cb2e088b858da3aa4f66e99e5c8444831e1701220142fa33640f41-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "500000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 81962827,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x15c1ee113cb1d5fe1093d4a45d7f276da473349023e1de51ace2aa2d27faf796",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x15c1ee113cb1d5fe1093d4a45d7f276da473349023e1de51ace2aa2d27faf796-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "880000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 81833253,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x23683155bebac309bac6cf932a55621adb393df306c58685a66410409428ba2f",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x23683155bebac309bac6cf932a55621adb393df306c58685a66410409428ba2f-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "750000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 83596403,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x33454a2e9f61a5679e981137f398972ee263d9ebd7813a09a6d85e481415e755",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x33454a2e9f61a5679e981137f398972ee263d9ebd7813a09a6d85e481415e755-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "100000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 83597516,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x36303441039c23917e6acf2dc32753815fb3d5f52b423075b5db7bc0be4ad991",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x36303441039c23917e6acf2dc32753815fb3d5f52b423075b5db7bc0be4ad991-OUT",
      "recipients": [
        "hxfc049b4ce60cc5968487d180d9c4263e63181739",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "1310000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 83289347,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x47fb7a49fa25cb2edb5425aa95ee11983605860e9abbe1037209bf10df0ed88d",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x47fb7a49fa25cb2edb5425aa95ee11983605860e9abbe1037209bf10df0ed88d-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "700000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 82173689,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x662110befe9b8925cec86f6c13d396f35cbabfe451f3302156eede5b4bb62dc4",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x662110befe9b8925cec86f6c13d396f35cbabfe451f3302156eede5b4bb62dc4-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "980000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 80919225,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x666196ba5b51d3d6beb18c25ba1e6ee5b8f72520678867f3e77ef446adcde235",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x666196ba5b51d3d6beb18c25ba1e6ee5b8f72520678867f3e77ef446adcde235-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "1234000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 82738092,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x794e6038769a07d9b016ca8f3869722dcff25be6dcf609c45772a154a4e6657b",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x794e6038769a07d9b016ca8f3869722dcff25be6dcf609c45772a154a4e6657b-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "1234000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 82737965,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x7fef9e6b4efe6f32fea8be4f3e605189572019dd6cbbbcca211212f5331a967f",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x7fef9e6b4efe6f32fea8be4f3e605189572019dd6cbbbcca211212f5331a967f-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "1234000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 81834658,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x804c4367a21da8ee5db04c3411b716efe130a69bf55ff105c67c16bc022145bf",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x804c4367a21da8ee5db04c3411b716efe130a69bf55ff105c67c16bc022145bf-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "870000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 81962885,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x90d3cf477c0b0e397a0b2b950fcb0baf9209907427b45409d4779a800dda71c4",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x90d3cf477c0b0e397a0b2b950fcb0baf9209907427b45409d4779a800dda71c4-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "880000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 82738311,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x93d741e63b803c6d6e5e0310e949260b85a5aa62827c93983946d35a2a88a9b6",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0x93d741e63b803c6d6e5e0310e949260b85a5aa62827c93983946d35a2a88a9b6-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "1200000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 82570455,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0xaa0ed3f7d60e5c4635c5afc8e3bee597dfab8632328a503b270041744b3fce78",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0xaa0ed3f7d60e5c4635c5afc8e3bee597dfab8632328a503b270041744b3fce78-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "1800000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 81316314,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0xaccf357bc8e9a8eb05ba16004590fdf36ffc79dd2518d168634afedf4978ebed",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0xaccf357bc8e9a8eb05ba16004590fdf36ffc79dd2518d168634afedf4978ebed-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "1464500000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 82173636,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0xbe34f687792b28441981a4cfaf40f85b0bd20a82129c217cf9659822a6fd1221",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0xbe34f687792b28441981a4cfaf40f85b0bd20a82129c217cf9659822a6fd1221-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "980000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 80823061,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0xce4c5afb7b422c5f754cd437271815e4830b2037b05a0bc9c80f7d86117859b5",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0xce4c5afb7b422c5f754cd437271815e4830b2037b05a0bc9c80f7d86117859b5-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "500000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 81316237,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0xda90a65b27bf6e0c58e00b9274d5fecdc2d637946d88334f6cc7429298492b23",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0xda90a65b27bf6e0c58e00b9274d5fecdc2d637946d88334f6cc7429298492b23-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "764500000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 80919366,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0xdc026e198438d6d7edcf25e92ac34027a04099bf741cee077a4df0fdb3aaf00a",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0xdc026e198438d6d7edcf25e92ac34027a04099bf741cee077a4df0fdb3aaf00a-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "700000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 82570388,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0xf6edc460f109a188bbbb7c0d539cf9153b348603e2281ff146a0f190e2140634",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0xf6edc460f109a188bbbb7c0d539cf9153b348603e2281ff146a0f190e2140634-IN",
      "recipients": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "senders": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "type": "IN",
      "value": "1800000000000000000",
    },
    {
      "accountId": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon",
      "blockHash": null,
      "blockHeight": 81834705,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0xf8d0bfbead922dd45527e8521145d9ed86a70d1396269ee5b80c47b5851807e6",
      "id": "js:2:icon:hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d:icon-0xf8d0bfbead922dd45527e8521145d9ed86a70d1396269ee5b80c47b5851807e6-OUT",
      "recipients": [
        "hx2dbaef576c73cca274db0b602831c2c6a9fc4bba",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "OUT",
      "value": "870000000000000000",
    },
  ],
  [
    {
      "accountId": "js:2:icon:hxfc049b4ce60cc5968487d180d9c4263e63181739:icon",
      "blockHash": null,
      "blockHeight": 83597516,
      "extra": {},
      "fee": "1250000000000000",
      "hasFailed": false,
      "hash": "0x36303441039c23917e6acf2dc32753815fb3d5f52b423075b5db7bc0be4ad991",
      "id": "js:2:icon:hxfc049b4ce60cc5968487d180d9c4263e63181739:icon-0x36303441039c23917e6acf2dc32753815fb3d5f52b423075b5db7bc0be4ad991-IN",
      "recipients": [
        "hxfc049b4ce60cc5968487d180d9c4263e63181739",
      ],
      "senders": [
        "hxcf4e2c94b52619f72aa4da13370aa6c1c3dad00d",
      ],
      "type": "IN",
      "value": "60000000000000",
    },
  ],
  [],
]
`;
